client/src/main.rs

```diff
@@ -109,97 +108,292 @@ pub async fn main() -> Result<(), Box<dyn Error>> {
+                return Err(indradb::Error::UuidTaken.compat())?;
+            }
+
+            println!("{:?}", vertex);
+        } else if let Some(matches) = matches.subcommand_matches("edge") {
+            let edge_key = build_edge_key(matches)?;
+            let res = client
+                .transaction()
+                .await
+                .map_err(|err| err.compat())?
+                .create_edge(&edge_key)
+                .await
+                .map_err(|err| err.compat())?;
+            if !res {
+                return Err(indradb::Error::VertexInvalid.compat())?;
+            }
+
+            println!("{:?}", edge_key);
+        } else if let Some(matches) = matches.subcommand_matches("vertex-property") {
+            let vertex_query = build_vertex_query(matches)?;
+            let property_name = matches.value_of("name").unwrap();
+            let property_value = serde_json::from_str(matches.value_of("value").unwrap())?;
+            client
+                .transaction()
+                .await
+                .map_err(|err| err.compat())?
+                .set_vertex_properties(VertexPropertyQuery::new(vertex_query, property_name), &property_value)
+                .await
+                .map_err(|err| err.compat())?;
+        } else if let Some(matches) = matches.subcommand_matches("edge-property") {
+            let property_name = matches.value_of("name").unwrap();
+            let property_value = serde_json::from_str(matches.value_of("value").unwrap())?;
+            let edge_query = build_edge_query(build_edge_key(matches)?)?;
+            client
+                .transaction()
+                .await
+                .map_err(|err| err.compat())?
+                .set_edge_properties(EdgePropertyQuery::new(edge_query, property_name), &property_value)
+                .await
+                .map_err(|err| err.compat())?;
         }
     } else if let Some(matches) = matches.subcommand_matches("count") {
-        if let Some(_matches) = matches.subcommand_matches("vertex") {
-            unimplemented!();
-        } else if let Some(_matches) = matches.subcommand_matches("edge") {
-            unimplemented!();
+        if let Some(_) = matches.subcommand_matches("vertex") {
+            let vertex_count = client
+                .transaction()
+                .await
+                .map_err(|err| err.compat())?
+                .get_vertex_count()
+                .await
+                .map_err(|err| err.compat())?;
+            println!("{}", vertex_count);
+        } else if let Some(matches) = matches.subcommand_matches("edge") {
+            let vertex_id = uuid::Uuid::parse_str(matches.value_of("id").unwrap()).map_err(|err| err.compat())?;
+            let edge_direction = match matches.value_of("inbound") {
+                Some(_) => indradb::EdgeDirection::Inbound,
+                None => indradb::EdgeDirection::Outbound,
+            };
+            let edge_type = match matches.value_of("type") {
```

Comment:
You can use [`Option::map`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map) to simplify this a bit

Reply:
I'm not sure how to cleanly bubble errors out of the closure without making the code equally verbose
For e.g if I write -
```rust
let edge_type = matches.value_of("type")
.map(|edge_type| indradb::Type::new(edge_type).map_err(|err| err.compat()));
``` 
the error thrown by the `new()` method needs to be handled in the outer function

Reply:
Ah right, yeah a match seems like the right move here